### DISCUSSION TITLE
Pin terraform modules to SHAs

### DISCUSSION THREAD

**Discussion by nitrocode** [renovatebot/renovate#31006] [Suggest an Idea]
### Tell us more.

Similar to docker digest pinning and gha digest pinning, it would be nice to pin terraform modules to a sha.

At the moment, the version argument doesn't support shas

Examples of unpinned shas

```hcl
module "ecr" {
  source  = "cloudposse/ecr/aws"
  version = "0.42.0"
}
```

```hcl
module "ecr" {
  source = "git::https://github.com/cloudposse/terraform-aws-ecr.git?ref=v0.42.0"
}
```

so converting the source from the terraform registry to a a git source sha would be the best way to do this.

Example of a pinned sha

```hcl
module "ecr" {
  source = "git::https://github.com/cloudposse/terraform-aws-ecr.git?ref=2896ea4fd21a41ebe86d6bed7f68735a51e83831
" # 0.42.0
}
```

Is this something renovate can be configured to do?

This may be affected by the issue/decision here https://github.com/opentofu/opentofu/issues/1942

Thanks!

**Comment by nitrocode** [comment]
cc @rarkins penny for your thoughts? This becomes more interesting after the recent tj-actions incident which was avoidable if the gha were pinned to a sha. The same issue can happen with terraform modules.

**Comment by rarkins** [comment]
This is only possible if we wrote our own parser for tf files to read in the annotations. When we use a HCL parser it will strip comments and return to use only the AST

**Reply by nitrocode** [reply]
I've been using this tool which is handy for making edits to terraform code and keeping comments.

https://github.com/minamijoyo/hcledit

If they can do it, I'm sure it's possible in renovate. Perhaps renovate can reuse the same library as hcledit.
